### ISSUE TITLE
Custom auth + interactive auth, switch user during login

### ISSUE THREAD

**Issue by ryancrunchi** [aws-amplify/aws-sdk-ios#2264]
**Describe the bug**
Let's say I'm using app with current username "123", the authentication flow requires me to login again with interactive auth and custom challenge.
I respond to challenge with details `USERNAME: 456`
These details are replaced by SDK internally by `USERNAME: 123` when sending auth request.
Specifically, the USERNAME is replaced by the userName of the AWSCognitoIdentityUser the interactive auth was initiated with (mostly `pool.currentUser()`).

**To Reproduce**
1. Login with a username
2. Make the credentials of this user so that it requires to auth again via interactive auth (expired tokens or delete user on another device)
3. Enter informations of another user during auth.

**Which AWS service(s) are affected?**
Cognito

**Expected behavior**
When switching user during interactive auth, use the informations given and update the currentUser in the pool with the last authenticated one.

**Screenshots**

**Environment(please complete the following information):**
 - SDK Version: 2.10.3
 - Dependency Manager: Cocoapods
 - Swift Version : 5

**Device Information (please complete the following information):**
 - Device: any
 - iOS Version: 13

**Additional context**

**Comment by ryancrunchi** [comment]
Technically, USERNAME request parameter is replaced in `addSecretHashDeviceKeyAndUsername` of the user instance.
I did not try with Android SDK, but by reading sources it seems the USERNAME parameter is not replaced.

**Comment by kneekey23** [comment]
hi @ryancrunchi can you possibly post a sample of code so I can see how you are using the Cognito SDK or are you using AWSMobileClient?

**Comment by ryancrunchi** [comment]
Hello,
I'm using 2 steps to authenticate.
First, user enter phone number and submit.
Then, user receive one-time sms code that he submits as answer.

Calling this function when I need user informations :
```
myUserPool.currentUser()?.getDetails().continueWith { ... }
```

which call interactive auth with AWSCognitoIdentityInteractiveAuthenticationDelegate
```
func startCustomAuthentication() -> AWSCognitoIdentityCustomAuthentication {
    return loginProvider
}
```
AWSCognitoIdentityCustomAuthentication
```
func getCustomChallengeDetails(_ authenticationInput: AWSCognitoIdentityCustomAuthenticationInput, customAuthCompletionSource: AWSTaskCompletionSource<AWSCognitoIdentityCustomChallengeDetails>) {
    switch step {
    case .phone:
        self.userNameChallengeCompletion = customAuthCompletionSource
        break

    case .code:
        self.codeChallengeCompletion = customAuthCompletionSource
        break
        
    default: break
    }
}
```
When user submits phone
```
        let details = AWSCognitoIdentityCustomChallengeDetails(challengeResponses: ["USERNAME": phoneInputValue])
        userNameChallengeCompletion?.set(result: details)
```
In the AWS SDK logs, the request parameter `USERNAME` is not the phone value given by the user.

**Comment by kneekey23** [comment]
hi @ryancrunchi can you walk me through the use case for switching users? The user that initiated the auth should be the one sent so I guess I am confused on the issue. It would be a security flaw if you could switch users at the point in time when you complete the challenge. I may be misunderstanding your use case. Can you please provide a bit more information about how you are trying to use this?

**Comment by ryancrunchi** [comment]
My app uses features users can play with without being connected.
And some other features that require account. For example if user tap profile menu, the interactive auth starts.
User connects as user1 username. Later, his session has expired and taps again profile menu. Interactive auth starts. User wants to connect to another account, as user2 username. The informations sent to cognito for initiateAuth are not user2, but user1

**Comment by kneekey23** [comment]
hi @ryancrunchi  when session has expired do you clear sessions and properly sign out a user?

**Comment by ryancrunchi** [comment]
No. How I know the session has expired before calling getSession and so interactive auth starts ?

**Comment by palpatim** [comment]
@ryancrunchi You have a few options for inspecting the user's state. The [`AWSCognitoIdentityUser`](https://aws-amplify.github.io/aws-sdk-ios/docs/reference/AWSCognitoIdentityProvider/Classes/AWSCognitoIdentityUser.html) object returned by `myUserPool.currentUser()` has a `signedIn` property that will be `true` if the user has an active session. You can also check `confirmedStatus`, which returns a [`AWSCognitoIdentityUserStatus`](https://aws-amplify.github.io/aws-sdk-ios/docs/reference/AWSCognitoIdentityProvider/Enums/AWSCognitoIdentityUserStatus.html) that indicates whether the user is confirmed. To clear locally cached data, including the "last known user", which is what is being sent in the second `initiateAuth` call as you describe above, you use `signOutAndClearLastKnownUser`.

**Comment by ryancrunchi** [comment]
ok
I see 2 edge cases here :
- currentUser is signed in (aka there is a refreshToken in keychain)
`isSigned` returns true, ok but I need fresh user data
I call `getDetails` which calls `getSession`

    - if `refreshToken` has expired or is not valid ?
The interactive auth starts, the `currentUser`'s `username` is used in place of the one I give.
`signOutAndClearLastKnownUser` does not clear the `currentUser`'s `username` the interactive auth was started with.

- currentUser is not signed in (there is no refreshToken in keychain)
calling `getSession` will obviously call the interactive auth.

    - What if the `currentUser`'s `username` is not nil (for whatever reason) ?
It will be used in place of the one I give, for the same reason of the above case.


To sum up, it is not a problem of token or session, but the `currentUser` instance with which you start the interactive auth. If it has, for any reason, a non-nil `username`, it will be used for `initiateAuth`, whatever else you give in the custom auth challenge response.
And nothing can change this `username` unless we change the `currentUser` instance, because `username` is readonly and set at initialization.
We can't either change currentUser instance because we have started interactive auth with this instance.

Sorry I can't find another way to explain. I hope I have not misunderstood the SDK, but it seems to be a case not taken into account.

**Comment by palpatim** [comment]
The common use case AWSMobileClient is trying to solve is automatically managing credentials for the lifecycle for a single user. We don't really consider use cases for automatically handling lifecycles where the user would change in the middle of a lifecycle--that's why we provide the `signOut` methods.

We don't support the ability to change the user being challenged in middle of the auth flow.

You might be able to mitigate this, for example, with a UI that shows the challenge response screen and a button to allow the user to completely sign out, as in:

```
Enter the verification code for <previous user>

Code: ____________
[Sign in]

[Not <previous user>? Click here to sign out and enter a new user.]
```

I hope this addresses the concern--I *think* I understand the use cases you're describing, but if I'm missing something please let me know.
